Support embeddable structs

dranikpg opened the pull request:
Hi! I've added support for embedded structs [as discussed in this issue](https://github.com/go-rel/rel/issues/241).

Both pointers and plain fields are supported.
```go
type Person struct {
	ID   int
	Name string
}

type Employee struct {
	Person
	Salary int
}
```

By default embedded field names are added without prefixes. Name prefixes can be set like for other fields:
```go
type Employee struct {
	Person   `db:"person_"`
	Salary int
}
```

If two fields have the same name, REL panics during construction:

### What I've changed

* `documentData` indices now store full field paths as []int
* `documentData`s are merged for embedded fields during creation
* Embedded pointers are initialized during document creation

### Tests

I've added tests for all main cases

dranikpg commented on association.go:
```diff
@@ -183,11 +185,11 @@ func newAssociation(rv reflect.Value, index int) Association {
 		key = associationKey{
 			rt:    rt,
-			index: index,
+			index: fmt.Sprint(index),
```
This could be replaced with a custom []int -> string function for better performance. Given, that those index slices are pretty short and contain small values, we can even hash them with a custom hashing function.

Fs02 replied:
how about printing the slice manually using `strconv` package
my concern is, `fmt.Sprint` is using interface that might be unnecessary here

dranikpg commented on document.go:
```diff
@@ -356,6 +356,20 @@ func (d Document) Flag(flag DocumentFlag) bool {
 	return d.data.flag.Is(flag)
 }
 
+func (d *documentData) mergeEmbedded(other documentData, pathPrefix int) {
```
This could be replaced by a function that takes two documents and produces one, but it has to show somehow that first is changed and no longer usable (because we don't want to fully copy slices)

Fs02 commented on document_test.go:
```diff
@@ -192,11 +192,37 @@ func TestDocument_Index(t *testing.T) {
+	}
+	var (
+		record = struct {
+			firstEmbedded
```
how about embedded pointer struct? is it supported?

Fs02 commented:
Thank you, this looking great 👍 

A few things to improve, can you add:
- Test case for getting value/primary key/primary field from embedded struct
- Test case for setting value to embedded struct
- How to handle embedded struct with duplicate fields

dranikpg commented:
I've added support for name prefixes and updated the description. 

> How to handle embedded struct with duplicate fields

This is now checked. I've also added some more tests.

---

`document.go` is now three lines over 500 and codeclimate is complaining. I'd suggest moving [pure reflect helper functions, like this](https://github.com/go-rel/rel/blob/1bbbcfdd01be1cf94c18d5d6b2c25c592bfacead/document.go#L245) to utils.

Fs02 commented on document.go:
```diff
@@ -356,6 +328,47 @@ func (d Document) Flag(flag DocumentFlag) bool {
+	for name, path := range other.index {
+		d.addFieldIndex(namePrefix+name, append([]int{indexPrefix}, path...))
+	}
+	appendWithPrefix := func(slice, newNames []string) []string {
```
maybe this function can be extracted?

dranikpg replied:
Should it really? It's here only to reduce repetitiveness by capturing the name prefix

Fs02 replied:
I prefer it that way rather then creating anonymous function
or alternatively this can be merged with `prefixFieldNames` as well 🤔

Fs02 commented:
> I've moved embedded pointer initialization to document creation

hmm, not sure about the correctness of this behavior.
for example if we query/preload a record, and it's not found, then we shouldn't modify the struct

dranikpg commented:
> > I've moved embedded pointer initialization to document creation
> 
> hmm, not sure about the correctness of this behavior. for example if we query/preload a record, and it's not found, then we shouldn't modify the struct

So all field reads/writes should be wrapped with pointer checks? I guess reads should init pointers, or else functions like PrimaryValue() could fail to return a value 😶

Fs02 commented:
> Not a problem, I just thought it's somewhat costly

yes right, but it's for the correctness, unless there's better  way to implement it 🤔

Fs02 commented:
> So all field reads/writes should be wrapped with pointer checks? Not a problem, I just thought it's somewhat costly.
I guess reads should init pointers, or else functions like PrimaryValue() could fail to return a value 😶

after giving it some thought and re-reading existing implementation, I think reads shouldn't initialize pointer
it should just returns nil if the embedded struct in path is nil

dranikpg commented on document.go:
```diff
@@ -477,6 +516,29 @@ func extractDocumentData(rt reflect.Type, skipAssoc bool) documentData {
 
+// Check pointers to embedded structs for index path
+// Returns false if pointers are invalid and init is false
+func checkPointersForIndices(rv reflect.Value, indices []int, init bool) bool {
```
deepsource is complaining because of the init parameter (countrol coupling). Should this really be split up? That'd end up with two almost identical functions

Fs02 replied:
I think we can ignore this

Fs02 commented on association.go:
```diff
@@ -131,7 +134,7 @@ func (a Association) ReferenceField() string {
 // ReferenceValue of the association.
 func (a Association) ReferenceValue() interface{} {
-	return indirectInterface(a.rv.Field(a.data.referenceIndex))
+	return indirectInterface(a.rv.FieldByIndex(a.data.referenceIndex))
```
isn't this and `ForeignValue` need to have nil embedded struct check as well?

Fs02 commented on association.go:
```diff
@@ -183,11 +186,21 @@ func newAssociation(rv reflect.Value, index int) Association {
 
-func extractAssociationData(rt reflect.Type, index int) associationData {
+// Encode index slice into single string
+func encodeIndices(indices []int) string {
```
let's move this to util.go?

Fs02 commented on document.go:
```diff
@@ -356,6 +337,52 @@ func (d Document) Flag(flag DocumentFlag) bool {
 }
 
+// Get field by index. Flag is false if this field is inaccessible
+func (d Document) fieldByIndex(index []int, init bool) (reflect.Value, bool) {
```
how about modifying original implementation of [FieldByIndex](https://cs.opensource.google/go/go/+/refs/tags/go1.17.7:src/reflect/value.go;bpv=1;bpt=1;l=1228?gsn=FieldByIndex&gs=kythe%3A%2F%2Fgolang.org%3Flang%3Dgo%3Fpath%3Dreflect%23method%2520Value.FieldByIndex) to add logic for lazy initialization?

I'm thinking something like this:
```golang
func reflectValueFieldByIndex(v reflect.Value, index []int, init bool) Value {
	if len(index) == 1 {
		return v.Field(index[0])
	}
	if v.Kind() != reflect.Struct {
		panic("rel: value must be a struct")
	}
	for i, x := range index {
		if i > 0 {
			if v.Kind() == Ptr && v.Type().Elem().Kind() == Struct {
				if v.IsNil() {
					if !init {
						return false
					}
					field.Set(reflect.New(v.Type().Elem()))
				}
				v = v.Elem()
			}
		}
		v = v.Field(x)
	}
	return v
}
```

dranikpg replied:
I've merged it with checkPointersForIndices, so yes, it doesn't walk down the types a second time

Fs02 commented on collection.go:
```diff
@@ -79,7 +79,7 @@ func (c Collection) PrimaryValues() []interface{} {
 			for j := 0; j < idxLen; j++ {
 				if item := c.rvIndex(j); item.IsValid() {
-					values = append(values, item.Field(index[i]).Interface())
+					values = append(values, item.FieldByIndex(index[i]).Interface())
```
this need nil embedded struct pointer check as well I think

Fs02 commented on util.go:
```diff
@@ -180,3 +181,43 @@ func fmtifaces(v []interface{}) string {
+	return sb.String()
+}
+
+// Get field by index and init pointers on path if flag is true
```
let's add some credit to original source here
something like: `modified from: https://cs.opensource.google/go/go/+/refs/tags/go1.17.7:src/reflect/value.go;l=1228-1245;bpv=1;bpt=1`

Fs02 commented on document.go:
```diff
@@ -184,9 +184,10 @@ func (d Document) Type(field string) (reflect.Type, bool) {
 // Value returns value of given field. if field does not exist, second returns value will be false.
 func (d Document) Value(field string) (interface{}, bool) {
 	if i, ok := d.data.index[field]; ok {
+		fv := reflectValueFieldByIndex(d.rv, i, false)
```
looks like we can group this together to line 189 again?

Fs02 reviewed (approved):
Thank you!!
just some more small comments, but overall already looking good 🎉

Fs02 commented:
Thanks for your contribution ❤️ 
if you have time, maybe you can also help updating the documentation?
https://github.com/go-rel/doc

dranikpg commented:
No problem, I'll have a look at it soon
